src/inspector_agent.cc

```diff
@@ -706,6 +773,74 @@ void Url(const FunctionCallbackInfo<Value>& args) {
   args.GetReturnValue().Set(OneByteString(env->isolate(), url.c_str()));
 }
 
+static void InvokeAsyncTaskFnWithId(void (Agent::*asyncTaskFn)(void*),
+                                    const FunctionCallbackInfo<Value>& args);
+
+static void* GetAsyncTask(int64_t asyncId);
+
+static void AsyncTaskScheduledWrapper(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  CHECK(args[0]->IsString());
+  Local<String> taskName = args[0].As<String>();
+  String::Value taskNameValue(taskName);
+  StringView taskNameView(*taskNameValue, taskNameValue.length());
+
+  CHECK(args[1]->IsNumber());
+  int64_t task_id = args[1]->IntegerValue(env->context()).FromJust();
+  void* task = GetAsyncTask(task_id);
+
+  CHECK(args[2]->IsBoolean());
+  bool recurring = args[2]->BooleanValue(env->context()).FromJust();
+
+  env->inspector_agent()->AsyncTaskScheduled(taskNameView, task, recurring);
+}
+
+static void AsyncTaskCanceledWrapper(const FunctionCallbackInfo<Value>& args) {
+  InvokeAsyncTaskFnWithId(&Agent::AsyncTaskCanceled, args);
+}
+
+static void AsyncTaskStartedWrapper(const FunctionCallbackInfo<Value>& args) {
+  InvokeAsyncTaskFnWithId(&Agent::AsyncTaskStarted, args);
+}
+
+static void AsyncTaskFinishedWrapper(const FunctionCallbackInfo<Value>& args) {
+  InvokeAsyncTaskFnWithId(&Agent::AsyncTaskFinished, args);
+}
+
+static void InvokeAsyncTaskFnWithId(void (Agent::*asyncTaskFn)(void*),
+                                    const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  CHECK(args[0]->IsNumber());
+  int64_t task_id = args[0]->IntegerValue(env->context()).FromJust();
+  (env->inspector_agent()->*asyncTaskFn)(GetAsyncTask(task_id));
+}
+
+static void* GetAsyncTask(int64_t asyncId) {
+  // The inspector assumes that when other clients use its asyncTask* API,
+  // they use real pointers, our at least something aligned like real pointer.
+  // In general it means that our task_id should always be even.
+  //
+  // On 32bit platforms, the 64bit asyncId would get truncated when converted
+  // to a 32bit pointer. However, the javascript part will never enable
+  // the async_hook on 32bit platforms, therefore the truncation will never
+  // happen in practice.
+  return reinterpret_cast<void*>(asyncId << 1);
+}
+
+static void RegisterAsyncHookWrapper(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  CHECK(args[0]->IsFunction());
+  v8::Local<v8::Function> enable_function = args[0].As<Function>();
+  env->inspector_agent()->RegisterAsyncHook(env->isolate(), enable_function);
+}
+
+static void EnabledGetter(Local<Name> property,
+                          const PropertyCallbackInfo<Value>& info) {
+  Environment* env = Environment::GetCurrent(info.GetIsolate());
+  info.GetReturnValue().Set(env->inspector_agent()->enabled());
```

Comment:
@TimothyGu  

> enabled() is actually not that good of a check. For instance, it is still true even after a process launched with `--inspect` has executed `inspector.close()`.

Good catch. I had a vague memory that there was a way how to disable inspector, but could not find it. Now I see there is `process._debugEnd()` that calls `env->inspector_agent()->Stop()` under the hood. IMO, we should be disabling the inspector async_hooks listener when the inspector was stopped - one more thing to fix before this patch is ready.

> If not, check out the C++ portions of TimothyGu/node@a9e18bf and the `IsListening()` function in TimothyGu/node@b01193f. `isListening() || isAttached()` should be what you are looking for.

Isn't it better to fix `inspector.enabled()` to correctly report `true` only when the agent is listening or attached and to report `false` after the inspector was stopped? I am thinking about adding a line `enabled_ = false` somewhere to the path executed by `InspectorAgent::Stop`. Thoughts?